build-system-abstraction.rst

```diff
@@ -0,0 +1,419 @@
+The build tool is then self-describing. This avoids having stale descriptions
+of build tools encoded into sdists that have been uploaded to PyPI [#pypi]_.
+
+The interface involves executing processes rather than loading Python modules,
+because tools like pip will often interact with many different versions of the
+same build tool during a single invocation, and in-process APIs are much
+harder to manage in that situation.
+
+Process interface
+-----------------
+
+Where a process needs to be run, we need to be able to assemble the subprocess
+call to make. For this we use a simple command string with in-Python variable
+interpolation. Only variables defined in this PEP will be honoured: the use of
+additional variables is an error. Additional variables can only be added with
+a schema version increase. Basic variable expressions such as ``${PYTHON}``,
+``${PYTHON:-python}``, ``${PYTHON:+PYTHON -m coverage}`` can be used. An
+implementation of this is in shellvars [#shellvars]_.
+
+Processes will be run with the current working directory set to the root of
+the source tree.
+
+Available variables
+-------------------
+
+PYTHON
+    The Python interpreter in use. This is important to enable calling things
+    which are just Python entry points::
+
+        ${PYTHON} -m foo
+
+OUTPUT_DIR
+    Where to create requested output. If not set, outputting to the current
+    working directory is appropriate. This is required by ``pip`` which needs
+    to have wheels output in a known location.
+
+pypa.json
+---------
+
+The yaml file has the following schema. Extra keys are ignored.
+
+schema
+    The version of the schema. This PEP defines version 1.
+    Defaults to 1
+
+bootstrap-requires
+    A list of dependency specifications that must be installed before
+    running the build tool. For instance, if using flit, then the requirements
+    might be::
+    
+        bootstrap-requires:
+          - flit
+
+build-tool
+    A command to run to query the build tool for its complete interface.
+    The build tool should output on stdout a build tool description JSON
+    document.  Stdin may not be read from, and stderr may be handled however
+    the calling process desires.
+
+build tool description
+----------------------
+
+The build tool description schema. Extra keys are ignored.
+
+schema
+    The version of the schema. This PEP defines version 1.
+    Defaults to 1
+
+build-requires
+    Command to run to query build requirements. Build requirements are
+    returned as a JSON document with one key ``build_requires`` consisting of
+    a list of dependency specifications. Additional keys must be ignored.
+
+metadata
+    Command to run to generate metadata for the project.  The metadata should
+    be output on stdout, stdin may not be consumed, and stderr handling is up
+    to the discretion of the calling process.  The build-requires for the
+    project will be present in the Python environment when the metadata
```

Comment:
Is this necessary? If I read this right it's very expensive. For example for projects that use the numpy C API, numpy itself has to be built/installed before you can get at the metadata for the project.